crates/matrix-sdk-ui/src/sync_service.rs

```diff
@@ -139,22 +143,17 @@ impl SyncTaskSupervisor {
                 if let Err(err) = encryption_sync.stop_sync() {
                     warn!(?report, "unable to stop encryption sync: {err:#}");
                 }
+
+                if report.has_expired {
+                    encryption_sync.expire_sync_session().await;
+                }
             }
 
             if let Err(err) = encryption_sync_task.await {
                 error!("when awaiting encryption sync: {err:#}");
             }
 
             if report.is_error {
-                if report.has_expired {
-                    if stop_room_list {
-                        room_list_service.expire_sync_session().await;
-                    }
-                    if stop_encryption {
-                        encryption_sync.expire_sync_session().await;
-                    }
-                }
-
```

Comment:
:warning: This commit is changing the semantics: before we'd only expire if we ran into an error, now it'll expire every time it's stopped. This should behave as it did beforehand, otherwise any small network cut will cause spurious sync expirations.

Reply:
As far as I understand, and please correct me if I'm wrong, `expired` can only be `true` if `error` is also true:

https://github.com/matrix-org/matrix-rust-sdk/blob/e151ea9664b7177cc73ffa6b3df0353750a5f185/crates/matrix-sdk-ui/src/sync_service.rs#L220

But that makes me think that these two booleans which aren't really independent should be replaced with an enum:

```rust
enum TerminationReason {
     Error,
     ExpiredAndError,
     Shrug,
}
```

This way we encode in the type that the fourth state (no error but expired) can't happen.

Reply:
Good point! Apart from the naming of `Shrug`, that sounds like a great thing to do (here or in a followup, as is preferred).

Reply:
I don't think I want to make this part of this PR.

Reply:
Your idea @poljar is nice and that's something I would enjoy reading.